apps/openmw/mwrender/sky.cpp

```diff
@@ -1445,6 +1448,72 @@ class WrapAroundOperator : public osgParticle::Operator
     }
 };
 
+class WorldCollisionOperator : public osgParticle::Operator
+{
+public:
+    WorldCollisionOperator(osg::Camera *camera, MWPhysics::PhysicsSystem* physicsSystem, float* rainCoverage): osgParticle::Operator()
+    {
+        mCamera = camera;
+        mPhysicsSystem = physicsSystem;
+        mRainCoverage = rainCoverage;
+    }
+
+    virtual osg::Object *cloneType() const override
+    {
+        return NULL;
+    }
+
+    virtual osg::Object *clone(const osg::CopyOp &op) const override
+    {
+        return NULL;
+    }
+
+    virtual void operate(osgParticle::Particle *P, double dt) override
+    {
+    }
+
+    virtual void operateParticles(osgParticle::ParticleSystem *ps, double dt) override
+    {
+        if(mPhysicsSystem->getRainCollisionEnabled()) {
+            int r = 0;
+            osg::Vec3 cameraPosition = getCameraPosition();
+
+            int numParticles = ps->numParticles();
+
+            for (int i = 0; i < numParticles; ++i) {
+                osgParticle::Particle *p = ps->getParticle(i);
+
+                osg::Vec3f pos = cameraPosition + p->getPosition();
+                osg::Vec3f extrapolatedOrigin = pos - (p->getVelocity() * 100);
+
+                MWPhysics::PhysicsSystem::RayResult result = mPhysicsSystem->castRay(extrapolatedOrigin, pos, MWWorld::ConstPtr(), std::vector<MWWorld::Ptr>(), MWPhysics::CollisionType_World);
```

Comment:
Keep in mind that not all game objects have precise collision mesh. Probably rendering raycasting will give better results.